### PR TITLE
Migrate to Construct 2.10

### PR DESCRIPTION
**Pull request by SupremeMortal**
Move away from an internal copy of Construct to the latest version in order to support new DWARFv5 data types.

### CODE REVIEW

**Comment by eliben** [comment]
Thanks, this is a very large change, PTAL at the CI test failures... Can you clarify how this is required to support DWARFv5 stuff?

**Comment by SupremeMortal** [comment]
Some of the form types mentioned [here](https://github.com/eliben/pyelftools/pull/363#issuecomment-847269798) require `UBInt24`. Although it would be possible to add these to the existing library, I noticed that it was on the TODO list to use the updated construct library so I thought I'd give it a shot.

Will investigate the CI failures.

**Comment by SupremeMortal** [comment]
CI and unit test issues have been resolved.

File: elftools/elf/sections.py [hunk 1]
```diff
@@ -480,38 +480,36 @@ def __init__(self, structs, stream):
...
 
             if self.tag != 'TAG_FILE':
                 self.extra = []
-                s_number = struct_parse(structs.Elf_uleb128('s_number'), stream)
+                s_number = struct_parse(structs.Elf_uleb128, stream)
 
                 while s_number != 0:
                     self.extra.append(s_number)
-                    s_number = struct_parse(structs.Elf_uleb128('s_number'),
+                    s_number = struct_parse(structs.Elf_uleb128,
```
Comment (eliben) [inline elftools/elf/sections.py:491]: Nit: fold into one line here too

File: elftools/elf/sections.py [hunk 2]
```diff
@@ -480,38 +480,36 @@ def __init__(self, structs, stream):
...
                                             stream)
 
         elif self.tag in ('TAG_CPU_RAW_NAME', 'TAG_CPU_NAME', 'TAG_CONFORMANCE'):
-            self.value = struct_parse(structs.Elf_ntbs('value',
-                                                       encoding='utf-8'),
+            self.value = struct_parse(structs.Elf_ntbs,
```
Comment (eliben) [inline elftools/elf/sections.py:495]: ditto here and below

File: elftools/elf/segments.py [hunk 3]
```diff
@@ -6,7 +6,9 @@
 # Eli Bendersky (eliben@gmail.com)
 # This code is in the public domain
 #-------------------------------------------------------------------------------
-from ..construct import CString
+from construct import CString
+
```
Comment (eliben) [inline elftools/elf/segments.py:10]: Nit: remove blank line

File: elftools/common/construct_utils.py [hunk 4]
```diff
@@ -101,12 +105,65 @@ class StreamOffset(Construct):
...
         return stream.tell()
-    def _build(self, obj, stream, context):
+    def _build(self, obj, stream, context, path):
         context[self.name] = stream.tell()
-    def _sizeof(self, context):
+    def _sizeof(self, context, path):
         return 0
+
+
+class EmbeddableStruct(Struct):
```
Comment (eliben) [inline elftools/common/construct_utils.py:118]: So the newer construct versions removed these? I wonder if they offer any alternatives -- writing these custom structs always felt like a kludge

File: elftools/dwarf/callframe.py [hunk 5]
```diff
@@ -277,17 +277,19 @@ def _parse_cie_augmentation(self, header, entry_structs):
...
+            b'z': 'length' / entry_structs.Dwarf_uleb128,
+            b'L': 'LSDA_encoding' / entry_structs.Dwarf_uint8,
+            b'R': 'FDE_encoding' / entry_structs.Dwarf_uint8,
             b'S': True,
-            b'P': Struct(
-                'personality',
-                entry_structs.Dwarf_uint8('encoding'),
-                Switch('function', lambda ctx: ctx.encoding & 0x0f, {
-                    enc: fld_cons('function')
+            b'P': 'personality' / Struct(
```
Comment (eliben) [inline elftools/dwarf/callframe.py:284]: Can you fix the formatting here a bit? It's hard to see that the following fields are inside the Struct
Maybe indent these parts more, or extract them into a separate var before `available_fields`?

**Review by eliben — CHANGES_REQUESTED** [review]
Overall very nice! Still thinking about this, left some initial comments for now

**Comment by eliben** [comment]
We're long overdue on this :-) One thing that is slightly annoying is that we're adding *more* special construct utils/parsers instead of removing existing ones!

@sevaa - please take a look as well, when you have some time.

@SupremeMortal did you have a chance to measure performance? I want to make sure we're not regressing here

File: elftools/common/construct_utils.py (hunk 4 above)
Comment (SupremeMortal) [reply elftools/common/construct_utils.py:74]: Unfortunately, there are no replacements as mentioned here https://github.com/construct/construct/issues/1027#issuecomment-1498657219

I did however manage to remove `RepeatUntilExcluding`

**Comment by SupremeMortal** [comment]
I have not measured performance properly yet, but on my system, the unit tests now take on average around 9.1 seconds compared to before when it took 8.3 seconds. There are some optimisations that can be made to improve it which I will investigate.

**Comment by eliben** [comment]
> I have not measured performance properly yet, but on my system, the unit tests now take on average around 9.1 seconds compared to before when it took 8.3 seconds. There are some optimisations that can be made to improve it which I will investigate.

Running `python3 test/all_tests.py` takes around half a minute (wallclock time) at least on my machine, so this can provide a bit more runway for comparison.

**Comment by sevaa** [comment]
On general grounds, I welcome not forking vendor packages gratutously :) That said, I'll take a closer look.

**Comment by SupremeMortal** [comment]
I'm unsure if the performance will be good enough. I tested with a 1.5GB ELF file with DWARF v5 debug, and it took 20 seconds with this PR compared to 10 seconds without which is quite a large regression in performance. 

I also tried optimising structs with the new lazy-loaded structures in construct 2.10, but it ended up being unusable for most of the structs in this codebase.

**Comment by eliben** [comment]
@SupremeMortal thanks for the update! This is indeed concerning. pyelftools doesn't excel at performance in general, so making it 2x slower for large inputs is a hard pill to swallow.

@arekbulski cc-ing you, given the discussion in #180

**Comment by sevaa** [comment]
So a new convention to learn. Doesn't look too terrible. It did seem unnatural to me all along that parsers needed to have a name attached to them just in case the parser is embedded in a struct.

As for compiling, we could start by storing the basic building blocks (trivial DWARF forms, nondynamic headers, etc) in compiled form and see where it takes us.

**Comment by arekbulski** [comment]
Compilation is a feature in Construct.

**Comment by sevaa** [comment]
I have some bandwidth currently, wanted to revisit this. There's been several releases of Construct since, maybe they've done something about performance - or maybe we could apply some tweaks on our side. I don't see @SupremeMortal publishing any changes since the initial commit - if they looked at the performance issues, it was not mature enough for publishing, and anyway, the fork needs to be merged with the recent changes to the master. If @SupremeMortal doesn't respond in a few days, I could fork the fork and sort of take over.

**Comment by sevaa** [comment]
FYI, for me, with this PR's fork and the latest construct, the performance difference with the master is not that great - the master is ~2% faster for a firehose parsing of DWARF CU/DIEs in a subset of our corpus. That's two percent, not times two,

**Comment by eliben** [comment]
> FYI, for me, with this PR's fork and the latest construct, the performance difference with the master is not that great - the master is ~2% faster for a firehose parsing of DWARF CU/DIEs in a subset of our corpus. That's two percent, not times two,

Interesting. I wonder what benchmark @SupremeMortal is alluding to in https://github.com/eliben/pyelftools/pull/477#issuecomment-1638443003

2% is definitely tolerable!

**Comment by eliben** [comment]
Superseded by #548

**Pull request closed without merging**